File: scripts/blast6fmt_addname.py
#! /usr/bin/env python3

import sys
import argparse, textwrap
#import taxoniq
import re
from Bio import Entrez
from retrying import retry

Entrez.email = '2812743l@alpha2.cvr.gla.ac.uk'

def argparse_line():
    parser = argparse.ArgumentParser(description='',
        formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument('-i', metavar='TXT',
        help='blast format 6 file', required=True)
    parser.add_argument('-o', metavar='TXT',
        help='add taxid and taxonomy name file', required=True)
    argv = vars(parser.parse_args())
    return argv

def taxoniq_get_taxid(accession_id):
    info = {}
    try:
        t1 = taxoniq.Taxon(accession_id=accession_id)
        for i in t1.lineage:
            rank = i.rank.name
            taxid = i.tax_id
            scientific_name = i.scientific_name
            info[rank] = {}
            info[rank]['taxid'] = str(taxid)
            info[rank]['name'] = scientific_name
    except KeyError:
        return info

    return info

@retry(stop_max_attempt_number=10)
def entrez_get_taxid(accession_id):
    info = {}
    handle = Entrez.efetch(db='Nucleotide', id=accession_id,
            rettype="gb", retmode="text")
    pattern = r'/db_xref="taxon:(\d+)"'
    for i in handle:
        match = re.search(pattern, i)
        if match:
            taxid = match.group(1)
            break
    if taxid:
        handle = Entrez.efetch(db = "Taxonomy", id = taxid)
        records = Entrez.read(handle)
        if 'LineageEx' in records[0]:
            for rank in records[0]['LineageEx']:
                info[rank['Rank']] = {}
                info[rank['Rank']]['taxid'] = rank['TaxId']
                info[rank['Rank']]['name'] = rank['ScientificName']
        else:
            return info
    else:
        return info

    return info

def add_name(input_file, output_file):
    total_info = {}
    output = open(output_file, 'w')
    output.write('Contig\tAccessonNumber\t'
                 'superkingdom\tsuperkingdom_name\t'
                 'phylum\tphylum_name\t'
                 'class\tclass_name\t'
                 'order\torder_name\t'
                 'family\tfamily_name\t'
                 'subfamily\tsubfamily_name\t'
                 'genus\tgenus_name\n')
    all_rank = ['superkingdom', 'phylum', 'class', 'order',
                'family', 'subfamily', 'genus']
    with open(input_file, 'r') as handle:
        for line in handle:
            newline = line.strip().split('\t')
            contig_id = newline[0]
            accession_id = newline[1]
            output_list = [contig_id, accession_id]
            if accession_id not in total_info:
                tax_info = entrez_get_taxid(accession_id)
                total_info[accession_id] = tax_info
            else:
                tax_info = total_info[accession_id]
            for rank in all_rank:
                if rank in tax_info:
                    taxid = tax_info[rank]['taxid']
                    name = tax_info[rank]['name']
                else:
                    taxid = 'N/A'
                    name = 'N/A'
                output_list.append(taxid)
                output_list.append(name)
            output.write('{0}\n'.format('\t'.join(output_list)))
    output.close()

def main():
    argv = argparse_line()
    add_name(argv['i'], argv['o'])

if __name__ == '__main__':
    main()

File: scripts/blast6fmt_addname.R
library(argparse)
library(taxonomizr)

parser <- argparse::ArgumentParser(description = "Add Taxonomy for blast6fmt")

parser$add_argument("-i", "--input", dest = "input_file", 
                    help = "blast6fmt file", type = "character")

parser$add_argument("-o", "--output", dest = "output_file", 
                    help = "add taxonnomy file", type = "character")

parser$add_argument("-db", "--database", dest = "database_file", 
                    help = "accessionTaxa.sql", type = "character")

args <- parser$parse_args()

if (is.null(args$input_file) || is.null(args$output_file) ||is.null(args$database_file)) {
  stop("Please use the -h parameter to see the help.")
}

my_dataframe <- data.frame(Contig=character(0),
                           AccessonNumber=character(0),
                           superkingdom=character(0),
                           superkingdom_name=character(0),
                           phylum=character(0),
                           phylum_name=character(0),
                           class=character(0),
                           class_name=character(0),
                           order=character(0),
                           order_name=character(0),
                           family=character(0),
                           family_name=character(0),
                           subfamily=character(0),
                           subfamily_name=character(0),
                           genus=character(0),
                           genus_name=character(0)
)

if (file.info(args$input_file)$size != 0){
  input_data <- read.csv(args$input_file, sep="\t", header=FALSE)
  database <- args$database_file
  
  for (i in 1:nrow(input_data)) {
      contig_name <- input_data[i, 1]
      accession <- input_data[i, 2]
      taxid <- accessionToTaxa(accession, database)
      lineage <- getTaxonomy(taxid, database, 
                             desiredTaxa = c("superkingdom", "phylum", "class", 
                                             "order", "family", "subfamily", 
                                             "genus"))
      lineage <- as.data.frame(lineage)
      ##################################################
      superkingdom_name <- lineage$superkingdom
      superkingdom <- getId(superkingdom_name, database)
      ##################################################
      phylum_name <- lineage$phylum
      phylum <- getId(phylum_name, database)
      ##################################################
      class_name <- lineage$class
      class <- getId(class_name, database)
      ##################################################
      order_name <- lineage$order
      order <- getId(order_name, database)
      ##################################################
      family_name <- lineage$family
      family <- getId(family_name, database)
      ##################################################
      subfamily_name <- lineage$subfamily
      subfamily <- getId(subfamily_name, database)
      ##################################################
      genus_name <- lineage$genus
      genus <- getId(genus_name, database)
      row <- data.frame(Contig = contig_name,
                        AccessonNumber = accession,
                        superkingdom = superkingdom,
                        superkingdom_name = superkingdom_name,
                        phylum = phylum,
                        phylum_name = phylum_name,
                        class = class,
                        class_name = class_name,
                        order = order,
                        order_name = order_name,
                        family = family,
                        family_name = family_name,
                        subfamily = subfamily,
                        subfamily_name = subfamily_name,
                        genus = genus,
                        genus_name = genus_name
      )
      my_dataframe <- rbind(my_dataframe, row)
  }
  my_dataframe[is.na(my_dataframe)] <- "N/A"
  write.table(my_dataframe, file = args$output_file, 
              sep='\t', quote = FALSE, row.names = FALSE)
} else {
    write.table(my_dataframe, file = args$output_file,
                sep='\t', quote = FALSE, row.names = FALSE)
}
